### ISSUE TITLE
Allgather gradients in Tensorflow distributed 

### ISSUE THREAD

**Issue by vineeths96** [tensorflow/tensorflow#41561]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.2
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
All gathering gradients in distributed Tensorflow would allow us to directly manipulate gradients in a much more efficient way. We can then encode (compress) gradients at each device(GPU) before cross-communication and then decode (or decompress) the gradients after all gathering. This would also allow us to combine gradients in a non-trivial way. This can allow researchers to develop efficient coding (or compression) schemes for reduced communication (something like Huffman coding). Currently, AFAIK only all reduce with SUM, MEAN option is supported.

**Will this change the current api? How?**

**Who will benefit with this feature?**
Researchers and developers working in coding techniques for distributed learning. 

**Any Other info.**
There seems to be some TODO mentioned over [here](https://github.com/tensorflow/tensorflow/blob/3b5b3cd58acb7b0192d43db444b8a4db66f5bae3/tensorflow/python/distribute/cross_device_ops.py#L928). But I was wondering if we can do the above in a much simpler way.
1. Encode (or compress) gradients in each device after the gradient calculation [here](https://github.com/tensorflow/tensorflow/blob/2f4444c1cff8ac07ab2c31d1ae23d23c66147126/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L434)
2. Let Tensorflow do all the operations across devices till all reduce.
3. Decode (or decompress) gradients just before the all reduce is called in the function [here](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/distribute/distribute_lib.py#L1905).

**Comment by nikitamaia** [comment]
Hi @vineeths96, thanks for providing this detailed feature request. This is an important feature and it has been requested before. Exposing all_gather at the strategy level is actively being worked on. I can definitely update this thread when there are any changes.

**Comment by vineeths96** [comment]
@nikitamaia Thank you for your reply.

I went through the latest RC of TF2.3 and did not see any advancements in this direction. I guess it would take quite some time before this rolls out public.

Do you have any alternative ideas about how this could be achieved (maybe playing with the source codes)?

**Comment by nikitamaia** [comment]
Sorry for my late response here, but hopefully the following help

This is a lower level function that will be used in the all_gather implementation: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/collective_ops.py#L74

And a function for testing purpose in limited situation: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/test_util.py#L33

**Comment by nikitamaia** [comment]
Updating and closing this thread because [`tf.distribute.Strategy.gather`](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy?version=nightly#gather) and [`tf.distribute.ReplicaContext.all_gather`](https://www.tensorflow.org/api_docs/python/tf/distribute/ReplicaContext?version=nightly#all_gather) have been added in nightly.
